x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java

```diff
@@ -103,6 +127,94 @@ public static String realmSslPrefix(RealmConfig.RealmIdentifier identifier) {
         }).collect(Collectors.toMap(Tuple::v1, Tuple::v2));
     }
 
+    /**
+     * Returns the domain name that the given realm is assigned to.
+     * Assumes {@code #verifyRealmNameToDomainNameAssociation} successfully verified the configuration.
+     */
+    public static @Nullable String getDomainForRealm(Settings globalSettings, RealmConfig.RealmIdentifier realmIdentifier) {
+        // TODO reserved realm settings need to be pulled into core
+        if (realmIdentifier.equals(new RealmConfig.RealmIdentifier("reserved", "reserved"))) {
+            // reserved realm cannot be ascribed to any domain
+            return null;
+        }
+        // file and native realms can be referred to by their default names too
+        for (String domainName : DOMAIN_TO_REALM_ASSOC_SETTING.getNamespaces(globalSettings)) {
+            Setting<List<String>> realmsByDomainSetting = DOMAIN_TO_REALM_ASSOC_SETTING.getConcreteSettingForNamespace(domainName);
+            for (String realmName : realmsByDomainSetting.get(globalSettings)) {
```

Comment:
After some more thoughts, I still think it's better if this method is not `static` but an instance method. We don't need a new class for it, it can just be a method of the existing Realms class. A few reasons:
* The comment says `Assumes {@code #verifyRealmNameToDomainNameAssociation} successfully verified the configuration.`. It means this method has a precondition that must be satisfied before invoking it. That's basically asking for some kinda of instantiation where object is properly prepared before answering questions.
* It simplies the method signature by dropping the `Settings` parameter.
* `Realms` knows about `List<Realm>` which is more efficient in finding out the realm domain. This method is needed by the Get Profile API, concretely, in `ProfileService#getProfile(String uid, ...)`. A more efficient lookup helps.
* For realms that are disabled, `Realms` is already computing a `List<RealmConfig>`. It is now thrown away after use. But we can easily keep it around to answer queries for disabled realms.

Reply:
I have pushed https://github.com/elastic/elasticsearch/pull/81968/commits/8a8a13cab84d4074503067557c4b4fe353a16db1 .

I still would've liked the realm to have a domain attribute, even though that might not be used in most places. It feels more OOP that way.

Reply:
I have no issue for `Realm` to have a `domain` attribute. I actually think it's a good idea after our discussion. The comment was just about whether this method should be static and where it should be located. Sorry if that wasn't clear enough.